admin/client/src/components/FormBuilder/FormBuilder.js

```diff
@@ -93,154 +87,110 @@ export class FormBuilderComponent extends SilverStripeComponent {
-    // match the detail view (admin/campaigns/set/:id/show).
-    // This causes the form builder to be unmounted during a fetch() call.
-    // this.setState({ isFetching: true });
-
     return this.formSchemaPromise;
   }
 
   /**
-   * Update handler passed down to each form field as a prop.
-   * Form fields call this method when their state changes.
-   *
-   * You can pass an optional callback as the third param. This can be used to
-   * implement custom behaviour. For example you can use `createFn` hook from
-   * your controller context like this.
-   *
-   * controller.js
-   * ...
-   * detailEditFormCreateFn(Component, props) {
-   *   const extendedProps = Object.assign({}, props, {
-   *     handleFieldUpdate: (event, updates) => {
-   *       props.handleFieldUpdate(event, updates, (formId, updateFieldAction) => {
-   *         const customUpdates = Object.assign({}, updates, {
-   *           value: someCustomParsing(updates.value),
-   *         };
-   *
-   *         updateFieldAction(formId, customUpdates);
-   *       });
-   *     },
-   *   });
-   *
-   *   return <Component {...extendedProps} />;
-   * }
-   * ...
-   *
-   * @param {object} event - Change event from the form field component.
-   * @param {object} updates - Values to set in state.
-   * @param {string} updates.id - Field ID. Required to identify the field in the store.
-   * @param {function} [fn] - Optional function for custom behaviour. See example in description.
-   */
-  handleFieldUpdate(event, updates, fn) {
-    if (typeof fn === 'function') {
-      fn(this.getFormId(), this.props.formActions.updateField);
-    } else {
-      this.props.formActions.updateField(this.getFormId(), updates);
-    }
-  }
-
-  /**
    * When the action is clicked on, records which action was clicked on
    * This can allow for preventing the submit action, such as a custom action for the button
    *
-   * @param event
-   * @param submitAction
+   * @param {Event} event
+   * @return {boolean}
    */
-  handleAction(event, submitAction) {
-    this.props.formActions.setSubmitAction(this.getFormId(), submitAction);
+  handleAction(event) {
+    // Custom handlers
     if (typeof this.props.handleAction === 'function') {
-      this.props.handleAction(event, submitAction, this.getFieldValues());
+      this.props.handleAction(event, this.getFieldValues());
     }
+
+    const name = event.currentTarget.name;
+
+    // Allow custom handlers to cancel event
+    if (!event.isPropagationStopped()) {
+      // Perform state update async to avoid component re-render *during* event bubbling,
+      // which causes the <form> event target to change, and the browser to abort bubbling.
+      // Needs to be shorter than the expected server response to avoid edge cases.
+      this.submittingAction = name;
+      this.submittingSetTimeout = setTimeout(() => this.setState({ submittingAction: name }), 50);
+    }
+
+    return true;
```

Comment:
Understandable, jQuery had an ancient convention where `return false` = `event.preventDefault() && event.stopPropagation()` in event handlers